### PR TITLE
fix!: update wrapper to make wrapping CJS sync

### PR DESCRIPTION
**Pull request by jmartin4563**
Function wrapper now splits logic between wrapping CommonJS functions and ES Module functions. Previously, all wrapping was async, which caused issues for CommonJS functions that relied on callbacks for "ending" functions. Now, CommonJS wrapping is sync, and ES Module wrapping is async, due to dynamic imports returning a promise. In order to properly split logic, we introduced a new environment variable called `NEW_RELIC_USE_ESM`. When set to true, the wrapper function will be in "ESM mode", and be async. The default behavior is "CJS mode", and will be sync.

**BREAKING CHANGE**: ES Module functions using the layer must now set the environment variable `NEW_RELIC_USE_ESM` to true to properly wrap said functions by using `import()`. Callback based functions are also no longer supported when using ES Modules due to the async nature of dynamic `import()`.

Additionally, I've retooled the existing tests as "unit" tests and introduced "integration" tests. The purpose of integration tests is to ensure that wrapping does not affect the return values of all the different ways customers can write lambdas. To accomplish this, I've introduced serverless-offline to the dev deps, as it's the next best way to emulate the full AWS runtime short of running directly in AWS.

Closes NR-121074

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -12,7 +12,7 @@ jobs:
       - setup_remote_docker
       - run:
           name: Node 14 unit tests
-          command: cd nodejs && npm run ciTest
```
Comment (jmartin4563) [inline .circleci/config.yml:15]: TODO: need to update the coverage stages.

Also: shouldn't we run the tests on branches?

File: README.md [hunk 2]
```diff
@@ -74,12 +74,15 @@ These steps will help you configure the layers correctly:
...
+AWS announced support for Node 18 as a Lambda runtime in late 2022, introducing `aws-sdk` version 3 for Node 18 only. This version of `aws-sdk` patches `NODE_PATH`, so ESM-supporting functions using `import` and top-level `await` should work as expected with Lambda Layer releases `v9.8.1.1` and above (Numerical layer versions vary by region and runtime). To configure the layer to leverage `import`, add the environment variable `NEW_RELIC_USE_ESM: true`.
+
+Note that if you use layer-installed instrumentation with the `NEW_RELIC_USE_ESM` environment variable, your function must use promises or async/await; callback based functions are not supported. The Node wrapper uses a [dynamic import](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/import) to attach to your function, which is an asynchronous operation. If you still need support for callback based functions, you will have to use the CommonJS based wrapper, which can be done by removing the `NEW_RELIC_USE_ESM` environment variable.
```
Comment (jmartin4563) [inline README.md:85]: TODO: need to go through docs-website to see if there are other places we should document this

File: nodejs/test/unit/esmHandler.tap.js [hunk 3]
```diff
@@ -0,0 +1,44 @@
+'use strict'
```
Comment (jmartin4563) [inline nodejs/test/unit/esmHandler.tap.js:1]: TODO: need to log tickets to improve the unit test coverage

Comment (jmartin4563) [reply nodejs/test/unit/esmHandler.tap.js:1]: https://issues.newrelic.com/browse/NEWRELIC-9305

File: nodejs/test/integration/esm/index.tap.js [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+    {
+      description:
+        'when handler returns a callback but defines a callback parameter',
+      expected: 'Hello Promise!',
+      path: '/dev/promise-with-defined-callback-handler',
+      status: 200,
+    },
+    {
+      description:
+        'when handler throws an expection in promise should return 502',
```
Comment (mrickard) [inline nodejs/test/integration/esm/index.tap.js:77]:
```suggestion
        'when handler throws an exception in promise should return 502',
```

File: nodejs/test/integration/esm/index.tap.js [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+      path: '/dev/throw-exception-in-promise-handler',
+      status: 502,
+    },
+    {
+      description:
+        'when handler throws an expection before calling callback should return 502',
```
Comment (mrickard) [inline nodejs/test/integration/esm/index.tap.js:83]:
```suggestion
        'when handler throws an exception before calling callback should return 502',
```

File: nodejs/test/integration/cjs/index.tap.js [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+      description: 'when handler uses an async function',
+      expected: 'foo',
+      path: '/dev/async-function-handler',
+      status: 200,
+    },
+    // NOTE: mix and matching of callbacks and promises is not recommended,
+    // nonetheless, we test some of the behaviour to match AWS execution precedence
+    {
+      description:
+        'when handler returns a callback but defines a callback parameter',
```
Comment (mrickard) [inline nodejs/test/integration/cjs/index.tap.js:69]:
```suggestion
        'when handler returns a promise but defines a callback parameter',
```

File: nodejs/test/integration/cjs/index.tap.js [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+      expected: 'Hello Promise!',
+      path: '/dev/promise-with-defined-callback-handler',
+      status: 200,
+    },
+    {
+      description:
+        'when handler throws an expection in promise should return 502',
```
Comment (mrickard) [inline nodejs/test/integration/cjs/index.tap.js:76]:
```suggestion
        'when handler throws an exception in promise should return 502',
```

File: nodejs/test/integration/cjs/index.tap.js [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+      path: '/dev/throw-exception-in-promise-handler',
+      status: 502,
+    },
+    {
+      description:
+        'when handler throws an expection before calling callback should return 502',
```
Comment (mrickard) [inline nodejs/test/integration/cjs/index.tap.js:82]:
```suggestion
        'when handler throws an exception before calling callback should return 502',
```

File: nodejs/test/integration/esm/index.tap.js [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+      description: 'when handler uses an async function',
+      expected: 'foo',
+      path: '/dev/async-function-handler',
+      status: 200,
+    },
+    // NOTE: mix and matching of callbacks and promises is not recommended,
+    // nonetheless, we test some of the behaviour to match AWS execution precedence
+    {
+      description:
+        'when handler returns a callback but defines a callback parameter',
```
Comment (mrickard) [inline nodejs/test/integration/esm/index.tap.js:70]:
```suggestion
        'when handler returns a promise but defines a callback parameter',
```

File: nodejs/test/integration/serverless.yml [hunk 10]
```diff
@@ -0,0 +1,226 @@
...
+    default: false
+    cjs: false
+    esm: true
+
+plugins:
+  - serverless-offline
+
+provider:
+  name: aws
+  runtime: nodejs18.x
```
Comment (mrickard) [inline nodejs/test/integration/serverless.yml:15]: While it's potentially complicating things, I'd wonder what results we'd get under other runtimes as well. We've encountered some language differences between runtimes before.

Which takes precedence when we're running in CircleCI jobs? The CircleCI environment or the `serverless-offline` provider runtime?

Comment (mrickard) [reply nodejs/test/integration/serverless.yml:15]: (We could probably just update this line to use an env var from the CircleCI job, to make sure.)

File: .circleci/config.yml (hunk 1 above)
Comment (mrickard) [reply .circleci/config.yml:15]: Agreed that we should run tests on branches.

**Comment by mrickard** [comment]
This is great--particularly how we're now able to test a much more comprehensive array of Lambda use cases!

File: nodejs/test/unit/cjsHandler.tap.js [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+const utils = require('@newrelic/test-utilities')
+
+tap.test('Layer Handler - CJS Function', (t) => {
+  t.autoend()
+  let handler
+  let helper
+  let originalEnv
+
+  t.beforeEach(() => {
+    originalEnv = process.env
```
Comment (mrickard) [inline nodejs/test/unit/cjsHandler.tap.js:14]:
```suggestion
    originalEnv = {...process.env}
```

File: nodejs/test/unit/cjsHandler.tap.js [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+
+    const newrelic = helper.getAgentApi() 
+
+    ;({ handler } = proxyquire('../../index', {
+      'newrelic': newrelic
+    }))
+  })
+
+  t.afterEach(() => {
+    process.env = originalEnv
```
Comment (mrickard) [inline nodejs/test/unit/cjsHandler.tap.js:27]:
```suggestion
    process.env = {...originalEnv}
```

File: nodejs/test/unit/esmHandler.tap.js [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+const utils = require('@newrelic/test-utilities')
+
+tap.test('Layer Handler - ESM Function', (t) => {
+  t.autoend()
+  let handler
+  let helper
+  let originalEnv
+
+  t.beforeEach(() => {
+    originalEnv = process.env
```
Comment (mrickard) [inline nodejs/test/unit/esmHandler.tap.js:14]:
```suggestion
    originalEnv = {...process.env}
```

File: nodejs/test/unit/esmHandler.tap.js [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+
+    const newrelic = helper.getAgentApi() 
+
+    ;({ handler } = proxyquire('../../index', {
+      'newrelic': newrelic
+    }))
+  })
+
+  t.afterEach(() => {
+    process.env = originalEnv
```
Comment (mrickard) [inline nodejs/test/unit/esmHandler.tap.js:27]:
```suggestion
    process.env = {...originalEnv}
```

**Review by mrickard — APPROVED** [review]

File: README.md (hunk 2 above)
Comment (jmartin4563) [reply README.md:85]: https://github.com/newrelic/docs-website/pull/13665

**Pull request merged**